apps/ledger-live-mobile/src/screens/SyncOnboarding/EarlySecurityCheck.tsx

```diff
@@ -1,42 +1,404 @@
+  let genuineCheckUiStepStatus: UiCheckStatus = "inactive";
+  let firmwareUpdateUiStepStatus: UiCheckStatus = "inactive";
+
+  // Handles genuine check states logic (both check state and UI state)
+  if (currentStep === "genuine-check") {
+    if (genuineCheckStatus === "ongoing") {
+      genuineCheckUiStepStatus = "active";
+      currentDisplayedDrawer = "none";
+
+      // Updates the genuineCheckStatus
+      if (genuineCheckError) {
+        log("EarlySecurityCheck", "Failed to run genuine check:", genuineCheckError.message);
+        setGenuineCheckStatus("error");
+      } else if (genuineState === "genuine") {
+        setGenuineCheckStatus("completed");
+      } else if (genuineState === "non-genuine") {
+        log("EarlySecurityCheck", "Device not genuine");
+        setGenuineCheckStatus("non-genuine");
+      }
+
+      // Updates the displayed drawer and UI state
+      if (devicePermissionState === "unlock-needed") {
+        // As the PIN has not been set before the ESC, the "unlock-needed" happens if the device is powered off.
+        // But an error `CantOpenDevice` should be triggered quickly after.
+        log("EarlySecurityCheck", "Genuine check permission state set to unlock-needed");
+      } else if (devicePermissionState === "requested") {
+        currentDisplayedDrawer = "allow-manager";
+      } else if (devicePermissionState === "refused") {
+        currentDisplayedDrawer = "none";
+        genuineCheckUiStepStatus = "genuineCheckRefused";
+      }
+    } else if (genuineCheckStatus === "error") {
+      // Currently genuine check error or refused is handled in the same way. This can be changed in the future.
+      currentDisplayedDrawer = "genuine-check-error";
+    } else if (genuineCheckStatus === "non-genuine") {
+      currentDisplayedDrawer = "non-genuine";
+    }
+  }
+  // `currentStep` can be any value for those UI updates
+  if (genuineCheckStatus === "completed") {
+    genuineCheckUiStepStatus = "completed";
+  } else if (["error", "non-genuine"].includes(genuineCheckStatus)) {
+    genuineCheckUiStepStatus = "error";
+  }
+
+  // Handles firmware update check UI logic
+  if (currentStep === "firmware-update-check") {
+    if (firmwareUpdateCheckStatus === "ongoing") {
+      firmwareUpdateUiStepStatus = "active";
+      currentDisplayedDrawer = "none";
+
+      // Updates the firmwareUpdateCheckStatus
+      // If the current error triggered a retry attempt, does not display failure
+      if (
+        getLatestAvailableFirmwareError &&
+        !(
+          getLatestAvailableFirmwareError.type === "SharedError" &&
+          getLatestAvailableFirmwareError.retrying
+        )
+      ) {
+        log(
+          "EarlySecurityCheck",
+          "Failed to retrieve latest firmware version with error:",
+          getLatestAvailableFirmwareError.name,
+        );
+        setFirmwareUpdateCheckStatus("error");
```

Comment:
I thought this was not really allowed in react (updating state inside the render function) and thought you should instead use a `useEffect` but it seems like it's actually ok in some cases like that https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes
Makes me think I know nothing about good practices in React 🤔🙈

Reply:
Yeah it's actually from this that i got the idea